internal/slicer/slicer.go

```diff
@@ -146,7 +145,64 @@ func Run(options *RunOptions) (*Report, error) {
 		packages[slice.Package] = reader
 	}
 
-	globbedPaths := make(map[string][]string)
+	type untilPathStatus int
+	const (
+		unset untilPathStatus = iota
+		mutate
+		keep
+	)
+	untilPaths := map[string]untilPathStatus{}
+
+	// Creates the filesystem entry and adds it to the report.
+	create := func(extractInfos []deb.ExtractInfo, o *fsutil.CreateOptions) error {
+		entry, err := fsutil.Create(o)
+		if err != nil {
+			return err
+		}
+		// Content created was not listed in a slice because extractInfo
+		// is empty.
+		if len(extractInfos) == 0 {
+			return nil
+		}
+
+		relPath := filepath.Clean("/" + strings.TrimLeft(o.Path, targetDir))
+		if o.Mode.IsDir() {
+			relPath = relPath + "/"
+		}
+		listed := false
+		until := untilPaths[relPath]
+		for _, extractInfo := range extractInfos {
+			if extractInfo.Context == nil {
+				continue
+			}
+			s, ok := extractInfo.Context.(*setup.Slice)
+			if !ok {
+				panic(fmt.Errorf("internal error: invalid Context of type %T in extractInfo", extractInfo.Context))
+			}
+			pathInfo, ok := s.Contents[extractInfo.Path]
+			if !ok {
+				panic(fmt.Errorf("internal error: path %q not listed in slice", extractInfo.Path))
+			}
+			listed = true
+
+			if pathInfo.Until == setup.UntilMutate && until == unset {
+				until = mutate
+			} else if pathInfo.Until == setup.UntilNone {
+				until = keep
```

Comment:
I have made the changes to use `setup.Until*`, to remove the local type and to make the checks as simple as possible. However, I am understanding your suggestion correctly, using `<` to compare `setup.Until*` seems a little bit brittle because we would be comparing strings present in the YAML in lexicographic order, meaning when we introduce a new value for `until: ` we would have to take that into account.

I have instead kept the manual check but I am happy to discuss it further.

Lastly, when we do the refactor in https://github.com/canonical/chisel/pull/134#discussion_r1616153064 we can deduplicate the logic and create a wrapper type and/or a helper function. But for this PR, I am keeping it like this.